cudax/include/cuda/experimental/__hierarchy/hierarchy_dimensions.cuh

```diff
@@ -103,69 +125,76 @@ struct get_level_helper
+  detail::has_level_helper<QueryLevel, ::cuda::std::__remove_cvref_t<Hierarchy>>::value;
 
 template <typename QueryLevel, typename Hierarchy>
 _CCCL_INLINE_VAR constexpr bool has_level_or_unit =
-  detail::has_level_helper<QueryLevel, ::cuda::std::remove_cv_t<::cuda::std::remove_reference_t<Hierarchy>>>::value
-  || detail::has_unit<QueryLevel, ::cuda::std::remove_cv_t<::cuda::std::remove_reference_t<Hierarchy>>>::value;
+  detail::has_level_helper<QueryLevel, ::cuda::std::__remove_cvref_t<Hierarchy>>::value
+  || detail::has_unit<QueryLevel, ::cuda::std::__remove_cvref_t<Hierarchy>>::value;
 
 namespace detail
 {
 template <typename... Levels>
 struct can_stack_checker
 {
   template <typename... LevelsShifted>
-  static constexpr bool can_stack = (detail::can_stack_on_top<LevelsShifted, Levels> && ...);
+  using can_stack = ::cuda::std::__fold_and<detail::can_stack_on_top<LevelsShifted, Levels>...>;
 };
 
-template <typename LUnit, ::cuda::std::size_t... Ids, typename... Levels>
-constexpr auto hierarchy_dimensions_fragment_reversed(::cuda::std::index_sequence<Ids...>, const Levels&&... ls);
+template <typename LUnit, typename L1, typename... Levels>
+using __can_stack =
+  typename can_stack_checker<__level_type_of<L1>,
+                             __level_type_of<Levels>...>::template can_stack<__level_type_of<Levels>..., LUnit>;
+
+template <size_t... _Id>
+_CUDAX_API constexpr auto __reverse_indices(::cuda::std::index_sequence<_Id...>) noexcept
+{
+  return ::cuda::std::index_sequence<(sizeof...(_Id) - 1 - _Id)...>();
+}
 
-template <bool Reversed, typename LUnit, typename L1, typename... Levels>
-_CCCL_NODISCARD constexpr auto make_hierarchy_fragment_reversable(L1&& l1, Levels&&... ls) noexcept
+template <typename LUnit, bool Reversed = false>
+struct __make_hierarchy_fragment
 {
-  using checker = can_stack_checker<typename ::cuda::std::remove_reference_t<L1>::level_type,
-                                    typename ::cuda::std::remove_reference_t<Levels>::level_type...>;
-  constexpr bool can_stack =
-    checker::template can_stack<typename ::cuda::std::remove_reference_t<Levels>::level_type..., LUnit>;
-  static_assert(can_stack || !Reversed,
-                "Provided levels can't create a valid hierarchy when stacked in the provided order or reversed");
-  if constexpr (can_stack)
+  template <class Levels, size_t... _Ids>
+  _CCCL_NODISCARD _CUDAX_TRIVIAL_API static constexpr auto
+  __apply_reverse(const Levels& ls, ::cuda::std::index_sequence<_Ids...>) noexcept
   {
-    return hierarchy_dimensions_fragment(LUnit{}, ::cuda::std::forward<L1>(l1), ::cuda::std::forward<Levels>(ls)...);
+    return __make_hierarchy_fragment<LUnit, true>()(::cuda::std::get<_Ids>(ls)...);
   }
-  else
+
+  template <typename L1, typename... Levels>
+  _CCCL_NODISCARD _CUDAX_API constexpr auto operator()(const L1& l1, const Levels&... ls) const noexcept
```

Comment:
I think we can now drop `l1`? It was needed previously for the `can_stack`, but now I think it's always used along with the ls pack.

Reply:
`__can_stack` wants to compare the shifted list of levels. we need to separate out `L1` here so we can easily shift the levels.

Reply:
Hmm, its not as easy as I thought, we would also need to change `__can_stack` to not be an alias, not sure if its worth it.
Feel free to revert my change here